### PR TITLE
[Merged by Bors] - Handle Geth pre-EIP-155 block sync error condition

### PR DESCRIPTION
**Pull request by jmcph4**
## Issue Addressed

#2293 

## Proposed Changes

 - Modify the handler for the `eth_chainId` RPC (i.e., `get_chain_id`) to explicitly match against the Geth error string returned for pre-EIP-155 synced Geth nodes
 - ~~Add a new helper function, `rpc_error_msg`, to aid in the above point~~
 - Refactor `response_result` into `response_result_or_error` and patch reliant RPC handlers accordingly (thanks to @pawanjay176)

## Additional Info

Geth, as of Pangaea Expanse (v1.10.0), returns an explicit error when it is not synced past the EIP-155 block (2675000). Previously, Geth simply returned a chain ID of 0 (which was obviously much easier to handle on Lighthouse's part).

### CODE REVIEW

**Comment by jmcph4** [comment]
Marking this as draft currently for a few reasons:

 - Confirming CI is happy
 - Likely prettier Rust that could be used
 - First time contributing to Lighthouse

**Comment by jmcph4** [comment]
I've been considering whether or not to tidy this up a bit but I feel it's overengineering at this point. Likely out of scope for this PR, but should we add types for handling [JSON-RPC 2.0 errors](https://www.jsonrpc.org/specification#error_object)?

File: beacon_node/eth1/src/http.rs
Comment (pawanjay176) [inline beacon_node/eth1/src/http.rs]: Working on the error string here by manually splitting the json string seems a little convoluted. Instead you can directly work on the json object by modifying the `response_result` function to return the error string itself. Something like
```rust
/// Accepts an entire HTTP body (as a string) and returns either the `result` field or the `error['message']` field, as a serde `Value` .
fn response_result_or_error(response: &str) -> Result<Result<Value, Value>, String> {
    let json = serde_json::from_str::<Value>(&response)
        .map_err(|e| format!("Failed to parse response: {:?}", e))?;

    if let Some(error) = json.get("error").map(|e| e.get("message")).flatten() {
        Ok(Err(error.clone()))
    } else {
        let res = json
            .get("result")
            .cloned()
            .ok_or_else(|| "Failed to get result".to_string())?;
        Ok(Ok(res))
    }
}
```

Then in the `get_chain_id` function, you can check the error string from the above function.

**Review by pawanjay176 — CHANGES_REQUESTED** [review]
This is great work! I had a suggestion to reduce manual string operations on the json string. Let me know what you think :)

File: beacon_node/eth1/src/http.rs
Comment (jmcph4) [reply beacon_node/eth1/src/http.rs]: This is much more elegant, provided we're happy with modifying the other handlers to accommodate the changed function signature?

Comment (jmcph4) [reply beacon_node/eth1/src/http.rs]: Actually I think this is a pretty clean change, consult e77a41a.

File: beacon_node/eth1/src/http.rs
Comment (pawanjay176) [inline beacon_node/eth1/src/http.rs]: Can directly return the result here so you don't have to handle the outer error case separately
```suggestion
        response_result_or_error(&response_body)?;
```

**Review by pawanjay176 — APPROVED** [review]
Looks great. Just a minor nitpick :)

File: beacon_node/eth1/src/http.rs
Comment (jmcph4) [reply beacon_node/eth1/src/http.rs]: Thanks again for this! I should've checked the return types more thoroughly. Consult 867afdb.

File: beacon_node/eth1/src/http.rs
Comment (paulhauner) [inline beacon_node/eth1/src/http.rs]: It's nitpick, but comments in the LH codebase generally look like this:

```suggestion
    // Extract response text here.
```

File: beacon_node/eth1/src/http.rs
Comment (paulhauner) [inline beacon_node/eth1/src/http.rs]: As per previous comment nitpick.

File: beacon_node/eth1/src/http.rs
Comment (paulhauner) [inline beacon_node/eth1/src/http.rs]:
```suggestion
/// Accepts an entire HTTP body (as a string) and returns either the `result` field or the `error['message']` field, as a serde `Value`.
```

**Review by paulhauner — CHANGES_REQUESTED** [review]
Awesome, thanks for getting involved @jmcph4. Glad to have your contributions :rocket: 

I've added a couple of nitpicks and a suggestion about the `response_result_or_error`. I know it conflicts with @pawanjay176, so both of you please let me know if I'm missing the mark :)

File: beacon_node/eth1/src/http.rs
Comment (jmcph4) [reply beacon_node/eth1/src/http.rs]: Also consult 31be1ab.

File: beacon_node/eth1/src/http.rs
Comment (jmcph4) [reply beacon_node/eth1/src/http.rs]: Consult 31be1ab.

File: beacon_node/eth1/src/http.rs
Comment (jmcph4) [reply beacon_node/eth1/src/http.rs]: @paulhauner As always, nitpicks are welcome! These are all valid changes to want for the codebase. 

Currently waiting on CI, although I don't see what would regress it.

**Comment by paulhauner** [comment]
> I've added a couple of nitpicks and a suggestion about the response_result_or_error

It seems like my comment regarding this was lost :( I'll rewrite it in the coming day(s).

File: beacon_node/eth1/src/http.rs
Comment (paulhauner) [inline beacon_node/eth1/src/http.rs]: I'm not sure about this `Result<Result<Value, Value>, String>`, it feels like an anti-pattern and I think the present implementation will result in some less-than-ideal errors.

For example, if we get an `"error"` field on `fn get_deposit_logs_in_range` we'll log an error saying `"No result field was returned for deposit logs"`, when it would be ideal to indicate that we got an error and the value of it.

I'd suggest something like this:

```rust
#[derive(Debug)]
enum RpcError {
    NoResultField,
    Error(String)
}

fn response_result_or_error(response: &str) -> Result<Value, RpcError> {
    ...
}
```
And then in `fn get_deposit_logs_in_range` we can do:

```rust
    response_result_or_error(&response_body)
        .map_err(|e| format!("Unable to obtain deposit logs: {:?}", e)?
```

**Review by paulhauner — CHANGES_REQUESTED** [review]
Thanks for the speedy response to my other comments. Here's that comment that got lost along the way, apologies for the delay.

File: beacon_node/eth1/src/http.rs
Comment (jmcph4) [reply beacon_node/eth1/src/http.rs]: Making an explicit error type would be better, agreed.

While it might be out of scope for this PR specifically, I feel like this gets back to [my comment](https://github.com/sigp/lighthouse/pull/2304#issuecomment-815539934) about supporting the entirety of the set of JSON-RPC 2.0 error conditions? After all, it seems like that's what Geth is sending to us anyway.

Comment (paulhauner) [reply beacon_node/eth1/src/http.rs]: > I feel like this gets back to my comment about supporting the entirety of the set of JSON-RPC 2.0 error conditions? After all, it seems like that's what Geth is sending to us anyway.

I don't feel so strongly either way about parsing the `error` struct instead of flattening it to a `String`. AFAIK, ultimately we just flatten it and present it to the user as a `String` anyway, so I'm not sure there's a whole lot to gain.

Comment (jmcph4) [reply beacon_node/eth1/src/http.rs]: > AFAIK, ultimately we just flatten it and present it to the user as a String anyway, so I'm not sure there's a whole lot to gain.

I think we err on the side of simplicity and just pass the string on, so I'm inclined to agree 
here.

I'll work on your proposed changes now.

**Comment by jmcph4** [comment]
@paulhauner Sorry this latest iteration took so long, hopefully the current changes are what you had in mind.

It appears that CI jobs aren't running for this anymore as I'm a first-time contributor? Strange given they've run fine every other time, but it might be me misunderstanding whatever the GitHub UI is trying to tell me.

**Comment by paulhauner** [comment]
Hi @jmcph4, sorry for the delay here. We've got a big PR backlog at the moment with Altair and Rayonism running in parallel. We're making good progress though and are coming back to PR reviews.

Would you mind fixing the CI conflict here please? I approved your CI run :)

**Comment by jmcph4** [comment]
@paulhauner dbe15afe fixes my `make audit` locally, but it appears that CI is only running the CLA check again.

As for the linting failure, `make lint` works for me locally, but I'll continue investigating.

**Comment by paulhauner** [comment]
> As for the linting failure, `make lint` works for me locally, but I'll continue investigating.

This is usually fixed by a `rustup update`, might be worth a shot :)

**Comment by jmcph4** [comment]
@paulhauner CI should be clear now and provided you're happy with the actual changeset I think this might be good to go? Let me know if I've missed anything (sorry this slid out for so long).

File: beacon_node/eth1/src/http.rs
Comment (paulhauner) [inline beacon_node/eth1/src/http.rs]: We've lost the ability to see the exact JSON error here. Sometimes `serde` gives useful information that can narrow down the error. Making this `InvalidJson(String)` would ensure we retain the error.

File: beacon_node/eth1/src/http.rs
Comment (paulhauner) [inline beacon_node/eth1/src/http.rs]: We're suppressing errors here.

Previously, if the server returned an `error` in the response we'd exit early with `Err`. Now, we return `Ok(None)`, indicating the server successfully returned with no data.

File: beacon_node/eth1/src/http.rs
Comment (paulhauner) [inline beacon_node/eth1/src/http.rs]: This would probably be more helpful as somthing like: "No result field in JSON response"

File: beacon_node/eth1/src/http.rs
Comment (paulhauner) [inline beacon_node/eth1/src/http.rs]: When we use `ok()` here we're still flattening all the different error types into "No result field was returned for block".

So, if there's an `RpcError::Error(s)` we're still presenting it to the user as if there was no result field.

File: beacon_node/eth1/src/http.rs
Comment (paulhauner) [inline beacon_node/eth1/src/http.rs]: As above regarding flattening errors.

File: beacon_node/eth1/src/http.rs
Comment (paulhauner) [inline beacon_node/eth1/src/http.rs]: As above regarding flattening errors.

File: beacon_node/eth1/src/http.rs
Comment (paulhauner) [inline beacon_node/eth1/src/http.rs]: As above regarding flattening errors.

**Review by paulhauner — CHANGES_REQUESTED** [review]
Thanks for persisting with this and apologies for the slow turn-around times.

There's a few changes I'd like to request. In the interests of giving feedback, I left a comment for each change request. In the interests of saving your time (and allowing me to understand it better) I created a commit which contains my recommendations:

https://github.com/paulhauner/lighthouse/commit/384b35413826d4fdaf2ddab28b73223bd0f8b04a

Feel free to just cherry-pick this commit onto your branch, if you agree with them.

**Comment by jmcph4** [comment]
> Thanks for persisting with this and apologies for the slow turn-around times.
> 
> There's a few changes I'd like to request. In the interests of giving feedback, I left a comment for each change request. In the interests of saving your time (and allowing me to understand it better) I created a commit which contains my recommendations:
> 
> [paulhauner@384b354](https://github.com/paulhauner/lighthouse/commit/384b35413826d4fdaf2ddab28b73223bd0f8b04a)
> 
> Feel free to just cherry-pick this commit onto your branch, if you agree with them.

@paulhauner Thanks for the feedback, it's much appreciated. I've reviewed your commit and cherry-picked it into this branch as you suggested. Apologies that this wasn't as smooth as it could've been.

**Comment by paulhauner** [comment]
> Thanks for the feedback, it's much appreciated. I've reviewed your commit and cherry-picked it into this branch as you suggested. Apologies that this wasn't as smooth as it could've been.

Thanks for your contribution, it's much appreciated. It's great to see people digging in and fixing annoyances :heart: 

We're waiting to clear the v1.4.0 release. Once that's done, we can merge this and include it in v1.5.0 :)

**Review by paulhauner — APPROVED** [review]
Let's get this merged! Thanks again @jmcph4.

bors r+

**Pull request closed without merging**
